File: aksharallm/portal/static/js/longctx.js
/* The Context tab: how far the model reads, what would extend it, and whether it worked.
 *
 * Two halves with deliberately different costs, and the UI is shaped to say so:
 *
 *   - "What would extending it do?" is a GET against pure arithmetic. Nothing is loaded,
 *     nothing is written, so it re-runs on every change of the dropdowns. A reader should be
 *     able to understand the whole trade-off before spending anything.
 *   - "Measure it" starts a detached job, one at a time, exactly like the Quantize tab. The
 *     device is chosen by the server (CPU while a run is training) and the reason is shown
 *     rather than left implicit.
 *
 * The charts are hand-drawn SVG like everywhere else in this portal, and every one of them
 * has a table twin — the position curve is the whole result of this tab and it has to be
 * readable without colour.
 */
import { $, api, escHtml, flash, fmt, post } from './core.js';
import { registerTab } from './router.js';

const lc = { loaded: false, ckpt: null, timer: null, results: [], open: null };

function status(text, kind = '') {
  const el = $('#lc-status');
  el.textContent = text;
  el.className = `panel-note${kind ? ` ${kind}` : ''}`;
}

/* ---- the free half ---------------------------------------------------------------------- */

function renderWindow(cur) {
  if (!cur) { $('#lc-window').innerHTML = ''; return; }
  const ratio = cur.trained_window ? cur.addressable / cur.trained_window : 1;
  const bar = Math.min(100, 100 / Math.max(ratio, 1));
  $('#lc-window').innerHTML = `
    <div class="lc-bar" role="img" aria-label="trained ${cur.trained_window} of ${cur.addressable} addressable tokens">
      <div class="lc-bar-trained" style="width:${bar}%"></div>
    </div>
    <dl class="lc-facts">
      <div><dt>trained on</dt><dd>${fmt.int(cur.trained_window)} tokens</dd></div>
      <div><dt>can address</dt><dd>${fmt.int(cur.addressable)} tokens</dd></div>
      <div><dt>scaling</dt><dd>${escHtml(cur.scaling && cur.scaling.type || 'none')}${
        cur.scaling && cur.scaling.factor > 1 ? ` ×${cur.scaling.factor}` : ''}</dd></div>
      <div><dt>sliding window</dt><dd>${cur.window ? `${fmt.int(cur.window)} tokens${
        cur.sinks ? ` + ${cur.sinks} sinks` : ''}` : 'none — full attention'}</dd></div>
    </dl>
    <p class="hint">${cur.extended
      ? 'This checkpoint has already been extended. Anything below compares against the window '
        + 'its <em>weights</em> were trained on, not the one it can address.'
      : 'Everything past the trained window is where the cliff lives. Try a method below.'}</p>`;
}

async function loadPlan() {
  if (!lc.ckpt) return;
  const method = $('#lc-method').value;
  const factor = $('#lc-factor').value;
  try {
    const p = await api(`/api/longctx/plan?checkpoint=${encodeURIComponent(lc.ckpt)}`
      + `&method=${method}&factor=${factor}`);
    const rows = p.changes.map((c) => `<li>${escHtml(c)}</li>`).join('');
    $('#lc-plan').innerHTML = `
      <ul class="lc-changes">${rows}</ul>
      <p class="lc-weights"><strong>No weights change.</strong> RoPE has no parameters — the
        rotation angles are computed from the position, so this is a config edit. The file
        <code>extend</code> writes has byte-for-byte the same tensors in it.</p>`;
    $('#lc-advice').textContent = p.advice;
  } catch (err) {
    $('#lc-plan').innerHTML = `<p class="err">${escHtml(err.message)}</p>`;
    $('#lc-advice').textContent = '';
  }
}

/* ---- the measured half ------------------------------------------------------------------- */

async function start(kind) {
  const body = { kind, checkpoint: lc.ckpt };
  /* The kernel benchmark measures the card, not a checkpoint — no model, no options, and
   * no CPU fallback, which the server enforces and explains. */
  if (kind === 'flash') {
    try {
      const res = await post('/api/longctx/start', body);
      flash(`Started — ${res.why_device}. First run compiles the kernels; give it a minute.`, 'ok');
      poll(1500);
    } catch (err) { flash(err.message, 'error'); }
    return;
  }
  if (kind === 'needle') {
    body.trials = Number($('#lc-trials').value) || 3;
    body.lengths = [512, 1024, 2048];
  } else {
    body.length = Number($('#lc-len').value);
    body.windows = Number($('#lc-windows').value) || 8;
    body.bucket = Math.max(64, Math.round(Number($('#lc-len').value) / 8));
    if (kind === 'sweep') {
      body.factor = Number($('#lc-factor').value);
      body.methods = ['none', 'linear', 'ntk', 'yarn'];
    } else {
      const m = $('#lc-method').value;
      if (m !== 'none') { body.factor = Number($('#lc-factor').value); }
    }
  }
  try {
    const res = await post('/api/longctx/start', body);
    flash(`Started — ${res.why_device}.`, 'ok');
    poll(600);
  } catch (err) { flash(err.message, 'error'); }
}

async function poll(delay = 2000) {
  clearTimeout(lc.timer);
  try {
    const res = await api(`/api/longctx?checkpoint=${encodeURIComponent(lc.ckpt || '')}`);
    const job = res.job || {};
    $('#lc-log').hidden = !job.log;
    if (job.log) $('#lc-log').textContent = job.log;
    $('#lc-stop').hidden = !job.running;
    $('#lc-job-note').textContent = job.running
      ? `running (pid ${job.pid})${res.training ? ` — on the CPU, ${res.training} has the card` : ''}`
      : (res.training ? `${res.training} is training, so a measurement would run on the CPU` : '');
    renderResults(res.results || []);
    if (job.running) lc.timer = setTimeout(() => poll(), delay);
  } catch (err) { /* the tab may have been left; nothing to say */ }
}

/* ---- results ------------------------------------------------------------------------------ */

function renderResults(rows) {
  lc.results = rows;
  if (!rows.length) {
    $('#lc-results').innerHTML = '<p class="hint">Nothing measured yet.</p>';
    return;
  }
  const body = rows.map((r) => {
    let summary = '';
    if (r.kind === 'sweep') {
      const best = r.methods.slice().sort((a, b) => a.loss - b.loss)[0];
      summary = `${r.methods.length} methods at ${fmt.int(r.seq_len)} — best `
        + `<strong>${escHtml(best.method)}</strong> ${fmt.num(best.loss, 3)}`;
    } else if (r.kind === 'needle') {
      summary = `${fmt.num(r.accuracy * 100, 1)}% retrieved (chance ${fmt.num(r.chance * 100, 0)}%)`;
    } else {
      summary = `loss ${fmt.num(r.loss, 3)}`;
    }
    return `<tr data-name="${escHtml(r.name)}"><th>${escHtml(r.kind)}</th>`
      + `<td>${escHtml(r.checkpoint || '')}</td><td>${summary}</td>`
      + `<td class="lc-when">${fmt.ago(r.when)}</td></tr>`;
  }).join('');
  $('#lc-results').innerHTML =
    `<table class="lc-table"><thead><tr><th>what</th><th>model</th><th>result</th><th></th>`
    + `</tr></thead><tbody>${body}</tbody></table>`;
  for (const tr of $('#lc-results').querySelectorAll('tr[data-name]')) {
    tr.addEventListener('click', () => openResult(tr.dataset.name));
  }
}

async function openResult(name) {
  try {
    const blob = await api(`/api/longctx/result?name=${encodeURIComponent(name)}`);
    lc.open = blob;
    if (blob.rows) renderSweep(blob);
    else if (blob.grid) renderNeedle(blob);
    else renderCurve(blob, blob.seq_len);
  } catch (err) { flash(err.message, 'error'); }
}

/* The position curve. One line per method, a dashed rule where the trained window ends —
 * which is the only annotation that matters, because the whole point is what happens after
 * it. Losses, not perplexities: a collapsed bucket's perplexity is in the millions and would
 * flatten every other line onto the axis. */
/* The palette is deliberately two categorical hues (see css/base.css), and a sweep has four
 * lines. Rather than invent colours the design system does not have, `none` takes the
 * *status* colour it has earned — it is the failure case, not a peer — and the three working
 * methods are separated by the two series hues plus a dash. Distinguishable without colour,
 * which is the point of the table underneath either way. */
const STROKES = {
  none:    { stroke: 'var(--critical)', dash: '' },
  linear:  { stroke: 'var(--series-2)', dash: '' },
  ntk:     { stroke: 'var(--series-1)', dash: '' },
  yarn:    { stroke: 'var(--series-1)', dash: '6 3' },
  dynamic: { stroke: 'var(--series-2)', dash: '6 3' },
  loss:    { stroke: 'var(--series-1)', dash: '' },
};

function curveSvg(series, trained, seqLen) {
  const W = 720, H = 260;
  const pad = { l: 46, r: 12, t: 16, b: 34 };
  const all = series.flatMap((s) => s.points.map((p) => p.loss));
  const lo = Math.min(...all) * 0.95, hi = Math.max(...all) * 1.05;
  const x = (pos) => pad.l + (pos / seqLen) * (W - pad.l - pad.r);
  const y = (v) => H - pad.b - ((v - lo) / (hi - lo || 1)) * (H - pad.b - pad.t);
  const pen = (label) => STROKES[label] || STROKES.loss;

  const lines = series.map((s) => {
    const d = s.points.map((p, j) =>
      `${j ? 'L' : 'M'}${x(p.start).toFixed(1)},${y(p.loss).toFixed(1)}`).join('');
    const { stroke, dash } = pen(s.label);
    return `<path class="series-line" d="${d}" stroke="${stroke}"`
      + `${dash ? ` stroke-dasharray="${dash}"` : ''}/>`;
  }).join('');

  const legend = series.map((s) => {
    const { stroke, dash } = pen(s.label);
    return `<span class="legend-item"><span class="legend-swatch" style="background:${stroke}`
      + `${dash ? ';opacity:.65' : ''}"></span>${escHtml(s.label)}</span>`;
  }).join('');

  /* The only annotation worth drawing: everything interesting happens to the right of it. */
  const edge = trained && trained < seqLen
    ? `<line class="rule-line" x1="${x(trained)}" y1="${pad.t}" x2="${x(trained)}" `
      + `y2="${H - pad.b}"/><text class="rule-label" x="${x(trained) + 5}" `
      + `y="${pad.t + 10}">trained window ends</text>`
    : '';
  const ticks = [0, 0.25, 0.5, 0.75, 1].map((f) =>
    `<text class="axis-label" x="${x(f * seqLen)}" y="${H - 12}" text-anchor="middle">`
    + `${fmt.int(f * seqLen)}</text>`).join('');
  const yticks = [lo, (lo + hi) / 2, hi].map((v) =>
    `<line class="grid-line" x1="${pad.l}" x2="${W - pad.r}" y1="${y(v)}" y2="${y(v)}"/>`
    + `<text class="axis-label" x="${pad.l - 8}" y="${y(v) + 4}" text-anchor="end">`
    + `${fmt.num(v, 1)}</text>`).join('');

  return `<div class="legend">${legend}</div>`
    + `<div class="chart"><svg viewBox="0 0 ${W} ${H}" role="img" `
    + `aria-label="cross-entropy loss by token position; the table below has the numbers">`
    + `<g>${yticks}${ticks}${edge}${lines}</g></svg></div>`;
}

function renderCurve(blob, seqLen) {
  const series = [{ label: 'loss', points: blob.buckets }];
  $('#lc-detail').innerHTML = curveSvg(series, blob.trained, seqLen || blob.seq_len)
    + table([{ label: 'loss', points: blob.buckets }]);
}

function renderSweep(blob) {
  const series = blob.rows.map((r) => ({ label: r.method, points: r.curve.buckets }));
  const rows = blob.rows.map((r) => {
    const bs = r.curve.buckets;
    const ins = bs.filter((b) => b.end < blob.trained).map((b) => b.loss);
    const out = bs.filter((b) => b.start >= blob.trained).map((b) => b.loss);
    const mean = (a) => (a.length ? a.reduce((s, v) => s + v, 0) / a.length : null);
    return `<tr><th>${escHtml(r.method)}</th><td>${fmt.num(r.curve.loss, 3)}</td>`
      + `<td>${fmt.num(mean(ins), 3)}</td><td>${fmt.num(mean(out), 3)}</td>`
      + `<td>${r.cliff ? fmt.int(r.cliff.position) : '—'}</td></tr>`;
  }).join('');
  $('#lc-detail').innerHTML = curveSvg(series, blob.trained, blob.seq_len)
    + `<table class="lc-table"><thead><tr><th>method</th><th>overall</th>`
    + `<th>in-window</th><th>past it</th><th>cliff at</th></tr></thead>`
    + `<tbody>${rows}</tbody></table>`
    + `<p class="hint">Read the <em>in-window</em> column as carefully as the others: a method
       that buys range by damaging short contexts has not helped. A cliff of “—” is the
       result you want.</p>`;
}

function table(series) {
  const rows = series[0].points.map((p, i) =>
    `<tr><th>${fmt.int(p.start)}–${fmt.int(p.end)}</th>`
    + series.map((s) => `<td>${fmt.num(s.points[i].loss, 3)}</td>`).join('') + '</tr>').join('');
  return `<table class="lc-table"><thead><tr><th>positions</th>`
    + series.map((s) => `<th>${escHtml(s.label)}</th>`).join('')
    + `</tr></thead><tbody>${rows}</tbody></table>`;
}

/* The needle grid. Cells are shaded against the chance line rather than against zero — a
 * grid of 25%s from a four-way choice is a grid of nothing, and colouring it as "a quarter
 * of the way there" would be the single most misleading thing this tab could draw. */
function renderNeedle(blob) {
  const head = blob.lengths.map((n) => `<th>${fmt.int(n)}</th>`).join('');
  const rows = blob.grid.map((row) => {
    const cells = row.map((c) => {
      if (c.accuracy == null) return '<td>—</td>';
      const over = (c.accuracy - blob.chance) / (1 - blob.chance);
      const shade = Math.max(0, Math.min(1, over));
      return `<td class="lc-cell" style="--w:${shade.toFixed(2)}">`
        + `${fmt.num(c.accuracy * 100, 0)}%</td>`;
    }).join('');
    return `<tr><th>${fmt.num(row[0].depth * 100, 0)}%</th>${cells}</tr>`;
  }).join('');
  const acc = blob.accuracy, se = blob.stderr;
  const chancey = se && acc - 2 * se <= blob.chance;
  $('#lc-detail').innerHTML =
    `<table class="lc-table lc-grid"><thead><tr><th>depth</th>${head}</tr></thead>`
    + `<tbody>${rows}</tbody></table>`
    + `<p class="${chancey ? 'err' : 'hint'}">Overall ${fmt.num(acc * 100, 1)}%`
    + (se ? ` ± ${fmt.num(se * 100, 1)}%` : '')
    + ` against a chance line of ${fmt.num(blob.chance * 100, 0)}%. `
    + (chancey
      ? 'That is <strong>not distinguishable from guessing</strong> — which is a real result, '
        + 'not a broken test. Scaling makes distant positions <em>legible</em>; being able to '
        + '<em>use</em> them is a capability, and capabilities come from training.'
      : 'Shading is against the chance line, not against zero.')
    + '</p>';
}

/* ---- wiring ------------------------------------------------------------------------------- */

async function openLongCtxTab() {
  if (!lc.loaded) {
    const res = await api('/api/longctx');
    const sel = $('#lc-ckpt');
    sel.innerHTML = (res.checkpoints || [])
      .map((c) => `<option value="${escHtml(c.rel)}">${escHtml(c.rel)}</option>`).join('');
    lc.ckpt = sel.value;
    $('#lc-len').innerHTML = (res.lengths || [])
      .map((n) => `<option value="${n}"${n === 2048 ? ' selected' : ''}>${n}</option>`).join('');

    sel.addEventListener('change', async () => {
      lc.ckpt = sel.value;
      await refreshCurrent();
      await loadPlan();
    });
    for (const id of ['#lc-method', '#lc-factor']) {
      $(id).addEventListener('change', loadPlan);
    }
    $('#lc-run-curve').addEventListener('click', () => start('curve'));
    $('#lc-run-sweep').addEventListener('click', () => start('sweep'));
    $('#lc-run-needle').addEventListener('click', () => start('needle'));
    $('#lc-run-flash').addEventListener('click', () => start('flash'));
    $('#lc-stop').addEventListener('click', async () => {
      await post('/api/longctx/stop', {});
      flash('Stop requested.', 'ok');
      poll(600);
    });
    lc.loaded = true;
  }
  await refreshCurrent();
  await loadPlan();
  poll();
}

async function refreshCurrent() {
  if (!lc.ckpt) { status('no checkpoints found', 'err'); return; }
  try {
    const res = await api(`/api/longctx?checkpoint=${encodeURIComponent(lc.ckpt)}`);
    renderWindow(res.current);
    renderResults(res.results || []);
    status(res.training ? `${res.training} is training — measurements will run on the CPU` : '');
  } catch (err) { status(err.message, 'err'); }
}

registerTab('longctx', { open: openLongCtxTab, leave: () => clearTimeout(lc.timer) });

File: aksharallm/portal/longctx.py
"""The Context tab's back end: how far this model can read, and what would extend it.

Shape of this one, and why it differs from its neighbours
----------------------------------------------------------
`quantize`, `finetune`, `evals` and `synth` shell out to a CLI and stream a log, because
those jobs run for minutes to hours. `interp` runs inline on the resident model, because a
logit lens is one forward pass.

Long context sits between the two, and it is split accordingly:

* **`plan`** is pure arithmetic — it answers "what would extending this checkpoint change?"
  without loading a single weight. That is the panel a layman should meet first: the whole
  trade-off is legible before anything is spent.
* **`curve`, `sweep` and `needle`** are real forward passes over long windows, so they run
  as a **background job**, one at a time, exactly like the Quantize tab — and with the same
  device policy, because a 4,096-token forward holds half a gigabyte of logits and the
  training run that owns the card has about three.

Results are the JSON files `python -m aksharallm.longctx` already writes into
`logs/longctx/`, so the tab and the terminal read the same measurements and neither is the
authority. Nothing here is recomputed for display.

Read with: docs/19-long-context.md -- the chapter this implements; it ends with the order to
read these files in.
"""

from __future__ import annotations

import json
import os
import signal
import subprocess
import sys

from pathlib import Path

from ..infer.checkpoints import InferError
from ..longctx.extend import default_out_name, describe, plan_extension
from ..model.rope import METHODS

#: Where the CLI leaves its JSON. The tab reads this directory and nothing else.
RESULTS = "logs/longctx"

#: One job at a time, same contract as the Quantize tab.
PID_FILE = "logs/longctx/job.pid"
LOG_FILE = "logs/longctx/job.log"

#: Lengths offered in the UI. Powers of two so the "x the trained window" label is exact.
LENGTHS = (512, 1024, 2048, 4096, 8192)


class LongContext:
    """Everything the Context tab can ask for."""

    def __init__(self, store, runs=None, root: Path | None = None):
        self.store = store
        self.runs = runs
        self.root = Path(root) if root else Path.cwd()

    # ---- the free part: arithmetic, no weights ------------------------------------------

    def overview(self, ckpt_id: str | None = None) -> dict:
        """Checkpoints, their current window, and whether a job is running."""
        cks = [c.as_dict() for c in self.store.list()]
        out = {
            "checkpoints": cks,
            "methods": [m for m in METHODS],
            "lengths": list(LENGTHS),
            "job": self.status(),
            "results": self.results(),
            "training": self._training(),
        }
        if ckpt_id:
            out["current"] = self.describe_checkpoint(ckpt_id)
        return out

    def describe_checkpoint(self, ckpt_id: str) -> dict:
        """What this checkpoint's context looks like right now — from the metadata only.

        `CheckpointStore` reads a `.pt` header with `mmap=True`, so this costs milliseconds
        and never loads weights. It is what lets the tab render the moment it opens, even
        while a training run owns the card.
        """
        info = self.store.get(ckpt_id)
        if info is None or info.error:
            raise InferError(f"cannot read {ckpt_id!r}: {info.error if info else 'no such checkpoint'}")
        scaling = info.rope_scaling or {"type": "none", "factor": 1.0}
        return {
            "checkpoint": ckpt_id,
            "trained_window": info.trained_window,
            "addressable": info.max_seq_len,
            "scaling": scaling,
            "extended": scaling.get("type") not in (None, "none"),
            "window": info.attn_window,
            "sinks": info.attn_sinks,
        }

    def plan(self, ckpt_id: str, method: str, factor: float) -> dict:
        """What `extend` *would* do. Pure, instant, and safe to call on every keystroke."""
        info = self.store.get(ckpt_id)
        if info is None or info.max_seq_len is None:
            raise InferError(f"{ckpt_id!r} does not record a context window")
        # `plan_extension` only ever reads these three keys, and `Checkpoint` already has
        # them from the header — so planning stays free rather than opening a 3.6 GB file.
        before = {"max_seq_len": info.max_seq_len, "rope_scaling": info.rope_scaling,
                  "attn_window": info.attn_window, "attn_sinks": info.attn_sinks}
        after = plan_extension(before, method, float(factor))
        return {
            "before": {"max_seq_len": before.get("max_seq_len"),
                       "rope_scaling": before.get("rope_scaling")},
            "after": {"max_seq_len": after.get("max_seq_len"),
                      "rope_scaling": after.get("rope_scaling")},
            "changes": describe(before, after),
            "weights_change": False,   # stated explicitly; it is the surprising part
            "advice": self._advice(method, float(factor)),
            # Named from `default_out_name`, the same function the CLI writes with — a
            # confirmation dialog quoting a path that turns out not to be the one used is
            # worse than no dialog at all.
            "out_name": (None if method == "none"
                         else default_out_name(info.path, method, float(factor)).name),
            "exists": (False if method == "none" else
                       default_out_name(info.path, method, float(factor)).exists()),
        }

    @staticmethod
    def _advice(method: str, factor: float) -> str:
        """One sentence, from our own measurements. The layman's version of the sweep table."""
        if method == "none":
            return "No scaling — the model falls off a cliff one token past its trained window."
        if method == "linear":
            return ("Simple, and it damages short contexts badly: on our 300M it took the "
                    "in-window loss from 2.356 to 3.035 at 2x, and to 4.379 at 4x.")
        if method == "ntk":
            return ("Best value up to about 2x — on our 300M, doubling the context cost "
                    "0.009 nats in-window (2.356 → 2.365). Effectively free."
                    if factor <= 2 else
                    "Past ~3.5x NTK's tilt runs out and grows a cliff of its own — on our "
                    "300M at 4x, at position 3,584 (loss 2.895 against YaRN's 2.464). "
                    "Try YaRN.")
        if method == "yarn":
            return ("The one that holds up furthest. Our 300M extended 4x this way scores "
                    "92.5% on the needle test at 4,096 tokens — four times the window its "
                    "weights were trained on, with nothing retrained.")
        return ("Recomputes the factor from each input's length, so prompts inside the "
                "original window are bit-for-bit unscaled. Stateful — read doc 18 first.")

    # ---- the measured part: a background job ---------------------------------------------

    def start(self, kind: str, ckpt_id: str, **opts) -> dict:
        """Launch a measurement detached. One job at a time.

        `flash` is the odd one out: it benchmarks the *kernel* rather than a checkpoint, so
        it runs `aksharallm.model.flash` instead and takes no model at all. It lives on this
        tab because it answers the other half of the same question — reading further costs
        attention, and attention is quadratic in how far you read.
        """
        if kind not in ("curve", "sweep", "needle", "flash", "extend"):
            raise InferError(f"unknown measurement {kind!r}")
        if self.status()["running"]:
            raise InferError("a measurement is already running — stop it or wait")

        if kind == "flash":
            if self._training():
                raise InferError(
                    "the kernel benchmark needs the GPU, and a run is training on it. "
                    "This is the one measurement here with no CPU fallback — the whole "
                    "point of it is what the card does.")
            cmd = [sys.executable, "-m", "aksharallm.model.flash",
                   "--seqlens", "512", "1024", "2048", "4096"]
            return self._spawn(cmd, kind, "cuda", "the card is free")

        if kind == "extend":
            # Writing a 3.6 GB copy is disk work; the device flag is irrelevant and `cpu`
            # keeps it away from a card someone may be using. No `--out`: the CLI names the
            # file from `default_out_name`, so no path ever arrives from a browser.
            method = str(opts.get("method") or "yarn")
            if method == "none":
                raise InferError("'none' removes a scaling rather than writing an extension")
            cmd = [sys.executable, "-m", "aksharallm.longctx", "extend", ckpt_id,
                   "--device", "cpu", "--quiet",
                   "--method", method, "--factor", str(opts.get("factor") or 4.0)]
            return self._spawn(cmd, kind, "cpu", "writing a checkpoint is disk work")

        cmd = [sys.executable, "-m", "aksharallm.longctx", kind, ckpt_id, "--quiet"]
        # The device decision is made here rather than offered as a checkbox: a long
        # forward pass beside a training run is how you lose a six-day run, and the tab
        # says which one it picked and why rather than silently choosing.
        device = "cpu" if self._training() else "cuda"
        cmd += ["--device", device]
        if device == "cuda":
            cmd.append("--force-gpu")   # we already checked; skip the CLI's own guard

        for flag, key in (("--len", "length"), ("--factor", "factor"),
                          ("--windows", "windows"), ("--bucket", "bucket"),
                          ("--trials", "trials"), ("--window", "window"),
                          ("--sinks", "sinks")):
            if opts.get(key) is not None:
                cmd += [flag, str(opts[key])]
        if opts.get("methods"):
            cmd += ["--methods", *[str(m) for m in opts["methods"]]]
        if opts.get("lengths"):
            cmd += ["--lengths", *[str(n) for n in opts["lengths"]]]

        return self._spawn(cmd, kind, device,
                           "a run is training, so this is on the CPU and will be slow"
                           if device == "cpu" else "the card is free")

    def _spawn(self, cmd: list[str], kind: str, device: str, why: str) -> dict:
        log = self.root / LOG_FILE
        log.parent.mkdir(parents=True, exist_ok=True)
        handle = log.open("w")
        handle.write(f"$ {' '.join(cmd)}\n\n")
        handle.flush()
        proc = subprocess.Popen(cmd, cwd=self.root, stdout=handle,
                                stderr=subprocess.STDOUT, start_new_session=True)
        (self.root / PID_FILE).write_text(str(proc.pid))
        return {"running": True, "pid": proc.pid, "kind": kind, "device": device,
                "why_device": why}

    def stop(self) -> dict:
        st = self.status()
        if not st["running"]:
            return {"running": False, "stopped": False}
        try:
            os.kill(st["pid"], signal.SIGTERM)
        except OSError:
            pass
        return {"running": False, "stopped": True}

    def status(self) -> dict:
        pid_file = self.root / PID_FILE
        log = self.root / LOG_FILE
        tail = log.read_text()[-4000:] if log.exists() else ""
        if not pid_file.exists():
            return {"running": False, "pid": None, "log": tail}
        try:
            pid = int(pid_file.read_text().strip())
            os.kill(pid, 0)
        except (ValueError, OSError):
            pid_file.unlink(missing_ok=True)
            return {"running": False, "pid": None, "log": tail}
        return {"running": True, "pid": pid, "log": tail}

    # ---- results ---------------------------------------------------------------------------

    def results(self, limit: int = 30) -> list[dict]:
        """Every measurement on disk, newest first. Headline numbers only — the full curve
        is fetched by `result()` when a row is opened, because a sweep's JSON carries every
        bucket of every method and there is no reason to send four of them to draw a list."""
        folder = self.root / RESULTS
        if not folder.is_dir():
            return []
        rows = []
        for path in sorted(folder.glob("*.json"), key=lambda p: p.stat().st_mtime,
                           reverse=True)[:limit]:
            try:
                blob = json.loads(path.read_text())
            except (json.JSONDecodeError, OSError):
                continue
            rows.append(self._headline(path, blob))
        return rows

    def result(self, name: str) -> dict:
        """One measurement in full. `name` is a bare filename — never a path."""
        if "/" in name or "\\" in name or not name.endswith(".json"):
            raise InferError("not a result name")
        path = self.root / RESULTS / name
        if not path.is_file():
            raise InferError(f"no result {name!r}")
        return json.loads(path.read_text())

    @staticmethod
    def _headline(path: Path, blob: dict) -> dict:
        row = {"name": path.name, "when": int(path.stat().st_mtime),
               "checkpoint": blob.get("checkpoint")}
        if "rows" in blob:            # a sweep
            row["kind"] = "sweep"
            row["seq_len"] = blob.get("seq_len")
            row["trained"] = blob.get("trained")
            row["methods"] = [
                {"method": r["method"], "loss": r["curve"]["loss"],
                 "cliff": (r.get("cliff") or {}).get("position")}
                for r in blob["rows"]]
        elif "grid" in blob:          # a needle sweep
            row["kind"] = "needle"
            row["accuracy"] = blob.get("accuracy")
            row["chance"] = blob.get("chance")
            row["lengths"] = blob.get("lengths")
        else:
            row["kind"] = "curve"
            row["loss"] = blob.get("loss")
        return row

    # ---- device policy ------------------------------------------------------------------

    def _training(self) -> str | None:
        """The name of a run currently holding the card, or None."""
        folder = self.root / "checkpoints"
        if not folder.is_dir():
            return None
        for pid_file in folder.glob("*/train.pid"):
            try:
                os.kill(int(pid_file.read_text().strip()), 0)
                return pid_file.parent.name
            except (ValueError, OSError):
                continue
        return None
